### PR TITLE
add discrete VAE

### PR DESCRIPTION
**Pull request by hnyu**
Add a VAE with discrete latent embedding. For gradient backprop, we use the straight-through estimator or straight-through Gumbel-softmax estimator (lower-variance).

It can be difficult for the discrete latent embedding to encode precisions in the inputs, unless the number of categorical variables is large. However, this kind of embedding might be helpful for MI computation between the latent embedding space and some other variable because we can control its granularity. For example, input x and y are encoded into the same set of discrete variables unless they are very different.

### CODE REVIEW

File: alf/algorithms/vae.py [hunk 1]
```diff
@@ -179,3 +184,143 @@ def train_step(self, inputs, state=()):
...
+class DiscreteVAE(VariationalAutoEncoder):
+    r"""VAE with a discrete posterior distribution. The latent ``z`` might be
+    a single categorical variable or a vector of categorials. Because the
+    re-parameterization trick can no longer be applied to the discrete distribution,
+    we instead use the straight-through gradient estimator to train the encoder.
+
+    ::
+
+        Bengio et al., "Estimating or Propagating Gradients Through Stochastic
+        Neurons for Conditional Computation", 2013.
```
Comment (emailweixu) [inline alf/algorithms/vae.py:200]: You may consider Gumbel softmax instead of straight through estimator, which is shown to perform better in https://arxiv.org/pdf/1611.01144.pdf

File: alf/algorithms/vae.py [hunk 2]
```diff
@@ -179,3 +184,143 @@ def train_step(self, inputs, state=()):
...
+            z_logits = prior_z_logits + dist_utils.distributions_to_params(
+                z_dist)['logits']
+            z_dist_spec = dist_utils.extract_spec(z_dist)
+            z_dist = dist_utils.params_to_distributions({
+                'logits': z_logits
+            }, z_dist_spec)
+        else:
+            uniform_prob = 1. / self._n_categories
+            entropy = dist_utils.compute_entropy(z_dist)
+            kl_div_loss = -np.log(uniform_prob) * self._z_spec.numel - entropy
```
Comment (emailweixu) [inline alf/algorithms/vae.py]: should not multiply with `self._z_spec.numel`

File: alf/algorithms/vae.py [hunk 3]
```diff
@@ -179,3 +184,143 @@ def train_step(self, inputs, state=()):
...
+        assert z_spec.is_discrete
+
+        prior_z_network = None
+        if prior_encoder_cls is not None:
+            prior_encoder = prior_encoder_cls(
+                input_tensor_spec=prior_input_tensor_spec)
+            prior_proj_net = OnehotCategoricalProjectionNetwork(
+                input_size=prior_encoder.output_spec.numel, action_spec=z_spec)
+            prior_z_network = alf.nn.Sequential(prior_encoder, prior_proj_net)
+            logits_spec = TensorSpec((np.prod(prior_proj_net._output_shape), ))
```
Comment (emailweixu) [inline alf/algorithms/vae.py]: do you think the code will be cleaner if we simply require that encoder and prior_encoder output the logits directly? And it will be quite easy to support Gumbel softmax.

Comment (hnyu) [reply alf/algorithms/vae.py]: Gumbel softmax seems not what I need. For discrete VAEs we would expect the sampled latent code to be completely discrete (hard labels). Gumbel softmax will still produce soft labels due to using softmax (with a temperature) to replace argmax. 

image

Comment (emailweixu) [reply alf/algorithms/vae.py]: As in the paper, it can also be combined with the straight through trick

Comment (hnyu) [reply alf/algorithms/vae.py]: > As in the paper, it can also be combined with the straight through trick

I'm actually a bit confused about combining Gumbel softmax and straight-through. The Gumbel trick (with argmax) is basically equivalent to Categorical sampling. If eventually we'll use straight-through on the one-hot, why bother using Gumbel trick to replace the categorical sampling in the first place? 

image

Comment (hnyu) [reply alf/algorithms/vae.py:270]: We could in fact support Gumbel softmax for the latent code, but comment on it to be softly sparse (depending on the temperature) instead of hard onehot. For hard latent code, still use the current onehot straight-through, as Gumbel softmax+straight-through seems unnecessary to me.

Comment (hnyu) [reply alf/algorithms/vae.py]: never mind, I think I understand now that there is still some difference between the two and the paper claims ST gumbel softmax to have lower variance than plain ST.

File: alf/algorithms/vae.py (hunk 2 above)
Comment (hnyu) [reply alf/algorithms/vae.py]: > should not multiply with `self._z_spec.numel`

self._z_spec.numel is the number of categoricals. The total KLD should be summed over them. `-np.log(uniform_prob)` is the entropy of one categorical.

File: alf/algorithms/vae.py [hunk 4]
```diff
@@ -179,3 +185,179 @@ def train_step(self, inputs, state=()):
...
+                 mode: str = "st",
+                 gumbel_temp_scheduler: Scheduler = ConstantScheduler(1.),
+                 beta: float = 1.,
+                 target_kld_per_categorical: float = None,
+                 beta_optimizer: torch.optim.Optimizer = None,
+                 name: str = "DiscreteVAE"):
+        """
+        Args:
+            z_spec: a tensor spec for the discrete posterior. Regardless of its
+                rank, it will be converted to rank-1, representing a vector of
```
Comment (emailweixu) [inline alf/algorithms/vae.py]: rank-1 is ambiguous. It can be "rank minus one" or "rank one"

File: alf/algorithms/vae.py [hunk 5]
```diff
@@ -179,3 +185,179 @@ def train_step(self, inputs, state=()):
...
+            beta: the weight for KL-divergence
+            target_kld_per_categorical: if not None, then this will be used as the
+                target KLD *per Categorical* to automatically tune beta.
+            beta_optimizer: if not None, will be used to train beta.
+            name (str):
+        """
+        Algorithm.__init__(self, name=name)
+
+        assert z_spec.is_discrete
+        self._n_categories = int(z_spec.maximum - z_spec.minimum + 1)
```
Comment (emailweixu) [inline alf/algorithms/vae.py]: Since the output z is always a vector and bounded in [0, n_gategories), to avoid misuse, we can simply assert z_spec.ndim==1 and z_spec.minimum==0.

File: alf/algorithms/vae.py [hunk 6]
```diff
@@ -179,3 +185,179 @@ def train_step(self, inputs, state=()):
...
+            dtype=torch.float32)
+
+    def _kl_divergence(self, logits1, logits2=None):
+        if logits2 is None:
+            logits2 = torch.zeros_like(logits1)  # assume uniform
+        logits1 = torch.nn.functional.log_softmax(logits1, dim=-1)
+        logits2 = torch.nn.functional.log_softmax(logits2, dim=-1)
+        kld_loss = torch.nn.KLDivLoss(reduction='none', log_target=True)
+        # The expectation is over the target distribution
+        kld = kld_loss(input=logits2, target=logits1)
```
Comment (emailweixu) [inline alf/algorithms/vae.py]: can use nn.functional.kl_div() directly.

**Review by emailweixu — APPROVED** [review]
